### ISSUE TITLE
VirtualPage doesn't find correct templates

### ISSUE THREAD

**Issue by henrypenny** [silverstripe/silverstripe-cms#1336]
The getViewer function of VirtualPage does not pass in a dataRecord - this causes the default Page.ss to be returned in all cases instead of the template suitable to render the 'copied' page.

	public function getViewer($action) {
		$originalClass = get_class($this->CopyContentFrom());
		if ($originalClass == 'SiteTree') $name = 'Page_Controller';
		else $name = $originalClass."_Controller";
		$controller = new $name();
		return $controller->getViewer($action);
	}

It needs to look more like:

	public function getViewer($action) {
		$originalClass = get_class($this->CopyContentFrom());
		if ($originalClass == 'SiteTree') $name = 'Page_Controller';
		else $name = $originalClass."_Controller";
                // pass in dataRecord of correct type.
		$controller = new $name(new $originalClass());
		return $controller->getViewer($action);
	}

**Comment by henrypenny** [comment]
I have patched this problem in my current project by adding this constructor to my Page_Controller:

class Page_Controller extends ContentController {

    /**
     * Page_Controller constructor.
     * @param null $dataRecord
     *
     * Constructor is required to patch an issue with Virtual pages not finding the correct template when no dataRecord is passed
     */
    public function __construct($dataRecord = null)
    {
        if($dataRecord == null) {
            $dataClass = get_class($this);
            $dataClass = str_replace('_Controller', '', $dataClass);
            $dataRecord = new $dataClass();
        }
        parent::__construct($dataRecord);
    }

**Comment by tractorcow** [comment]
Wouldn't this result in rendering a page with all of it's fields nullified?

**Comment by tractorcow** [comment]
Maybe VirtualPage_Controller::getViewer should look like the below:

```php
public function getViewer($action) {
	$object = $this->CopyContentFrom();
	$controller = ModelAsController::controller_for($object, $action);
	return $controller->getViewer($action);
}
```

**Comment by henrypenny** [comment]
It looks like that Controller instance is only created to locate the template in this instance.

The main problem is this:

    class ContentController extends Controller {
        ...
	    public function __construct($dataRecord = null) {
		    if(!$dataRecord) {
			    $dataRecord = new Page();
                            ...

If you don't pass in a $dataRecord the controller assumes that its a Page instance. And you end up rendering the Page.ss template.

Passing an empty model of the correct type has solved the problem in my project without any side effects that I can see. Although I have not stepped through the code exhaustively to determine if that Controller instance is retained for any other purpose.

**Comment by henrypenny** [comment]
I'm not super familiar with the code base, so I would guess your fix will be more correct than mine.

**Comment by tractorcow** [comment]
I think the existing code simply hasn't been changed for a while... `ModelAsController::controller_for` is the new standard way of getting a controller for any object. :D

**Comment by oetiker** [comment]
Any progress on this ?

**Comment by dhensby** [comment]
Someone needs to PR @tractorcow's suggestion

**Comment by tractorcow** [comment]
Fixed with #1605
